CovertInfiltration/Src/CovertInfiltration/Classes/CI_XComGameState_CovertAction.uc

```diff
@@ -22,3 +22,135 @@ protected function bool DisplaySelectionPrompt()
+	if (StrategyMap != none && StrategyMap.m_eUIState != eSMS_Flight && !`HQPRES.ScreenStack.IsCurrentClass(class'UIAlert'))
+	{
+		if (!bCompleted)
+		{
+			// If the end date time has passed, this action has completed
+			if (bStarted && class'X2StrategyGameRulesetDataStructures'.static.LessThan(EndDateTime, GetCurrentTime()))
+			{
+				if (class'X2Helper_Infiltration'.static.IsInfiltrationAction(self))
+				{
+					`log("Spawning infiltration mission for" @ m_TemplateName,, 'CI');
+
+					MissionState = XComGameState_MissionSiteInfiltration(NewGameState.CreateNewStateObject(class'XComGameState_MissionSiteInfiltration'));
+					MissionState.SetupFromAction(NewGameState, self);
+				}
+				else
+				{
+					`log(m_TemplateName @ "finished, it was not an infiltration",, 'CI');
+
+					// Note: infiltrations cannot have risks, period
+					ApplyRisks(NewGameState);
+				}
+				
+				if (bAmbushed)
+				{
+					// Flag XComHQ as expecting an ambush, so we can ensure the Covert Action rewards are only granted after it is completed
+					XComHQ = XComGameState_HeadquartersXCom(NewGameState.ModifyStateObject(class'XComGameState_HeadquartersXCom', XComHQ.ObjectID));
+					XComHQ.bWaitingForChosenAmbush = true;
+				}
+				else
+				{
+					CompleteCovertAction(NewGameState);
+
+					if (MissionState != none)
+					{
+						// Do not show the CA report, the mission will show its screen instead
+						bNeedsActionCompletePopup = false;
+
+						// Remove the CA, the mission takes over from here
+						RemoveEntity(NewGameState);
+
+						// We need to be sure that the player still cannot access those soldiers in armory, or use them in any way
+						MissionState.SetSoldiersAsOnAction(NewGameState);
+					}
+				}
+
+				bCompleted = true;
+				bModified = true;
+			}
+		}
+		else if (bAmbushed && !XComHQ.bWaitingForChosenAmbush)
+		{
+			bAmbushed = false; // Turn off Ambush flag so we don't hit this code block more than once
+			
+			// If the mission was ambushed, rewards were not granted before the tactical battle, so give them here
+			CompleteCovertAction(NewGameState);
+			bModified = true;
+		}
+	}
+
+	return bModified;
+}
+
+function GiveRewards(XComGameState NewGameState)
+{
+	if (class'X2Helper_Infiltration'.static.IsInfiltrationAction(self))
+	{
+		// The reward is the mission, you greedy
+		return;
+	}
+
+	super.GiveRewards(NewGameState);
+}
+
+function RemoveEntity(XComGameState NewGameState)
+{
+	local XComGameState_ResistanceFaction FactionState;
+	local bool SubmitLocally;
+
+	if (NewGameState == None)
+	{
+		SubmitLocally = true;
+		NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("Covert Action Despawned");
+	}
+
+	// The only change: do not kick people from finished infiltration
+	if (bStarted && class'X2Helper_Infiltration'.static.IsInfiltrationAction(self))
+	{
+		// Do not remove people from slots - we will do it later
```

Comment:
Maybe clarify what "later" is and why it's guaranteed to happen (i.e. comment that `RemoveEntity` can only be called with an infiltration mission active due to condition <X>)?